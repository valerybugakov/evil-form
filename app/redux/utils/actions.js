/* flow */
import { createAction } from 'redux-act'
import { identity, noop } from 'lodash/fp'
import { fakeStore } from 'redux/utils/dispatch'

export const INIT = 'init'
export const SUCCESS = 'success'
export const FAILURE = 'failure'

const requestMetaReducer = (actionName, {
  serializer,
  deserializer,
  applySerializerToPayload,
  ...rest
}) => ((payload, resolve = noop, reject = noop) => ({
  resolve,
  reject,
  serializer,
  deserializer,
  type: actionName,
  requestStatus: INIT,
  applySerializerToPayload,
  requestOptions: typeof rest.requestOptions === 'function'
    ? rest.requestOptions(payload)
    : rest,
}))

const createRequestStatusAction = (actionName, status) => (
  createAction(`${actionName} ${status}`, identity, (_, payload) => ({
    type: actionName,
    requestStatus: status,
    ...payload,
  }))
)

export const createRequestActions = ({
  desc: actionName,
  ...metaData
}) => {
  const res = {
    id: actionName,
    [INIT]: createAction(
      `${actionName} ${INIT}`,
      identity,
      requestMetaReducer(actionName, metaData),
    ).assignTo(fakeStore),
  }

  return [SUCCESS, FAILURE].reduce((reqs, type) => {
    reqs[type] = createRequestStatusAction(actionName, type)
    return reqs
  }, res)
}

export const createActionBinded = (value, process, meta) => (
  createAction(value, process, meta).assignTo(fakeStore)
)

export const promisifyAction = (action, payload) => (
  new Promise((resolve, reject) => action(payload, resolve, reject))
)

export function actionsGenerator(actionConfig, path = []) {
  const minimumActionConfig = {
    desc: path.join('.'),
  }

  // Create ordinary action if null provided as config
  if (actionConfig === null) {
    return createActionBinded(minimumActionConfig.desc)
  }

  // Create requestActions if requestOptions or url provided in config
  if (actionConfig.requestOptions || actionConfig.url) {
    return createRequestActions({
      ...minimumActionConfig,
      ...actionConfig,
    })
  }

  // Recursively call actionsGenerator on all config keys
  if (typeof actionConfig === 'object') {
    return Object.entries(actionConfig).reduce((acc, [key, value]) => {
      acc[key] = actionsGenerator(value, [...path, key])
      return acc
    }, {})
  }

  return actionConfig
}
