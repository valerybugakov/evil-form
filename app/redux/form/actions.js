import { actionsGenerator } from 'redux/utils'

export default actionsGenerator({
  test: null,
})

