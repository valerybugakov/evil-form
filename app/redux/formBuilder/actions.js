import { createActionBinded } from 'redux/utils'

export const saveForm = createActionBinded('Save form init')
export const saveFormSuccess = createActionBinded('Save form success')
export const saveFormFailure = createActionBinded('Save form failure')
