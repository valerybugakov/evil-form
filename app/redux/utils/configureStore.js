import { autoRehydrate } from 'redux-persist'
import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import { fakeStore } from 'redux/utils/dispatch'

const sagaMiddleware = createSagaMiddleware()

export function configureStore(
  reducer,
  initialState = {},
  rootSaga
) {
  const enhancers = [
    applyMiddleware(sagaMiddleware),
    autoRehydrate(),
  ]

  const composeEnhancers = (
    process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      : compose
  )

  const store = createStore(
    reducer,
    initialState,
    composeEnhancers(...enhancers),
  )

  // Assign real dispatch to `fakeStore` which was used to prebind actions
  fakeStore.dispatch = store.dispatch

  if (rootSaga) {
    sagaMiddleware.run(rootSaga)
  }

  return store
}
