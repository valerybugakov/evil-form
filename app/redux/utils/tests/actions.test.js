import { createReducer } from 'redux-act'
import { configureStore } from '../configureStore'
import { createActionBinded, promisifyAction } from '../actions'

describe('action utils', () => {
  describe('createActionBinded', () => {
    it('creates action with already binded dispatch', () => {
      const testAction = createActionBinded()

      const reducer = createReducer({
        [testAction]: (state, payload) => ({
          value: state.value + payload,
        }),
      })

      const store = configureStore(reducer, { value: 0 })

      testAction(7)
      expect(store.getState()).toEqual({ value: 7 })
    })
  })

  describe('promisifyAction', () => {
    it('wraps action into a promise with resolve fn in meta', async () => {
      const testAction = createActionBinded()

      const reducer = createReducer({
        [testAction]: (state, payload, { resolve }) => {
          const value = state.value + payload
          resolve(value)
          return { value }
        },
      })

      const store = configureStore(reducer, { value: 0 })
      const value = await promisifyAction(testAction, 7)

      expect(value).toEqual(7)
      expect(store.getState()).toEqual({ value: 7 })
    })
  })
})
