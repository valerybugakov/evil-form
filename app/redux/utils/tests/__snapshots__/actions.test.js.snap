// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action utils createActionBinded creates action with already binded dispatch 1`] = `
Object {
  "value": 7,
}
`;

exports[`action utils promisifyAction wraps action into a promise with resolve fn in meta 1`] = `
Object {
  "value": 7,
}
`;
